server/graphql/v2/mutation/PaymentMethodMutations.js

```diff
@@ -1,61 +1,76 @@
+  args: {
+    paymentMethod: {
+      type: PaymentMethodCreateInput,
+      description: 'A Payment Method to add to an Account',
+      deprecationReason: '2020-08-24: Please use creditCardInfo',
     },
-    async resolve(_, args) {
-      const account = await fetchAccountWithReference(args.account);
-      const collective = await models.Collective.findByPk(account.id);
-      if (!collective) {
-        throw Error('This collective does not exist');
-      }
-
-      const { paymentMethod } = args;
+    creditCardInfo: {
+      type: CreditCardCreateInput,
+      description: 'The credit card info',
+    },
+    name: {
+      type: GraphQLString,
+      description: 'Name associated to this credit card',
+    },
+    isSavedForLater: {
+      type: GraphQLBoolean,
+      description: 'Wether this payment method should be saved for future payments',
+    },
+    account: {
+      type: new GraphQLNonNull(AccountReferenceInput),
+      description: 'Account to add Payment Method to',
+    },
+  },
+  async resolve(_, args, req) {
+    const collective = await fetchAccountWithReference(args.account, { throwIfMissing: true });
+    if (!req.remoteUser.isAdminOfCollective(collective)) {
+      throw new Forbidden(`Must be an admin of ${collective.name}`);
+    }
 
-      const newPaymentMethodData = {
-        ...paymentMethod,
-        service: 'stripe',
-        type: 'creditcard',
-        CreatedByUserId: account.CreatedByUserId,
-        currency: account.currency,
-        saved: true,
-        CollectiveId: account.id,
-      };
+    const newPaymentMethodData = {
+      service: 'stripe',
+      type: 'creditcard',
+      name: args.name,
+      CreatedByUserId: req.remoteUser.id,
+      currency: collective.currency,
+      saved: args.isSavedForLater,
+      CollectiveId: collective.id,
+    };
 
-      let pm = await models.PaymentMethod.create(newPaymentMethodData);
+    if (args.creditCardInfo) {
+      const data = pick(args.creditCardInfo, ['brand', 'country', 'expMonth', 'expYear', 'fullName', 'funding', 'zip']);
+      Object.assign(newPaymentMethodData, { token: args.creditCardInfo.token, data });
+    } else if (args.paymentMethod) {
+      Object.assign(newPaymentMethodData, pick(args.paymentMethod, ['data', 'name', 'token']));
+    } else {
+      throw new ValidationFailed('Either creditCardInfo or paymentMethod must be provided');
+    }
 
-      try {
-        pm = await setupCreditCard(pm, {
-          collective,
-          user: account,
-        });
-      } catch (error) {
-        if (!error.stripeResponse) {
-          throw error;
-        }
+    // TODO Add support for 3D secure
+    const pm = await models.PaymentMethod.create(newPaymentMethodData);
+    return setupCreditCard(pm, { collective, user: req.remoteUser });
+  },
+};
 
-        pm.stripeError = {
-          message: error.message,
-          response: error.stripeResponse,
-        };
-      }
```

Comment:
Removed that as we're not supporting `stripeError` on GQLV2 at the moment, which is going to throw an error to users if stripe asks for 3D secure. We should obviously iterate on that, but in the mean time is that ok to merge @znarf?

Reply:
What about editing a payment method in Recurring Contributions? Does that mean Confirmation ("3Dsecure") is not supported there?

Reply:
Yes, it's not supported currently. If we face a 3D secure issue, it will just be ignored an the update will succeed. But now that I think about it, https://github.com/opencollective/opencollective-api/blob/c27ed4d726d5794c7a3ccc1a47a017ee1a571c80/templates/emails/payment.creditcard.confirmation.hbs should be sent once the payment fails no (in such case, we should continue to ignore this error)?